File: frontend/src/types/game.ts
export type CardType = 'attack' | 'skill' | 'ability'
export type CardRarity = 'common' | 'uncommon' | 'rare'
export type EnemyIntentType = 'attack' | 'defend' | 'buff' | 'debuff' | 'unknown'

export interface Card {
  id: string
  name: string
  description: string
  type: CardType
  rarity: CardRarity
  cost: number
  damage?: number
  block?: number
  effects?: CardEffect[]
  upgraded?: boolean
}

export interface CardEffect {
  type: string
  value?: number
  target?: 'self' | 'enemy' | 'all_enemies'
}

export interface Player {
  hp: number
  maxHp: number
  block: number
  energy: number
  maxEnergy: number
  gold: number
  buffs: Buff[]
  debuffs: Debuff[]
}

export interface Enemy {
  id: string
  name: string
  hp: number
  maxHp: number
  block: number
  intent: EnemyIntent
  buffs: Buff[]
  debuffs: Debuff[]
}

export interface EnemyIntent {
  type: EnemyIntentType
  value?: number
}

export interface Buff {
  id: string
  name: string
  stacks: number
}

export interface Debuff {
  id: string
  name: string
  stacks: number
}

export interface Deck {
  drawPile: Card[]
  hand: Card[]
  discardPile: Card[]
  exhaustPile: Card[]
}

export type GamePhase = 'map' | 'battle' | 'reward' | 'shop' | 'rest' | 'event' | 'victory' | 'game_over'

export interface GameState {
  gameId: string
  phase: GamePhase
  turn: number
  player: Player
  enemies: Enemy[]
  deck: Deck
  map: GameMap
  currentFloor: number
  currentRoom: number
  rngSeed: string
  pendingRewards: Card[]
}

export interface GameMap {
  floors: MapFloor[]
}

export interface MapFloor {
  rooms: MapRoom[]
}

export interface MapRoom {
  id: string
  type: RoomType
  completed: boolean
  connections: string[]
}

export type RoomType = 'battle' | 'elite' | 'boss' | 'rest' | 'shop' | 'event' | 'unknown'

export interface GameAction {
  type: 'play_card' | 'end_turn' | 'select_reward' | 'select_map_room' | 'rest_action' | 'shop_action'
  payload?: Record<string, unknown>
}

export interface GameStartRequest {
  character?: string
  seed?: string
  ascension?: number
}

export interface GameStartResponse {
  gameId: string
  gameState: GameState
}

export interface RewardCard {
  card: Card
  selected: boolean
}

export interface RewardOffer {
  cards: RewardCard[]
  goldReward: number
  canSkip: boolean
}

export interface SelectRewardRequest {
  cardIndex?: number
  skip?: boolean
}

export interface RestActionRequest {
  actionType: 'heal' | 'upgrade'
}

export interface ShopItem {
  card: Card
  price: number
  itemType: 'card_add' | 'card_remove'
}

export interface ShopOffer {
  items: ShopItem[]
  removePrice: number
}

export interface ShopActionRequest {
  itemIndex: number
  action: 'buy' | 'remove'
}

File: backend/schemas/game.py
from enum import Enum
from typing import List, Optional
from pydantic import BaseModel


class CardType(str, Enum):
    ATTACK = "attack"
    SKILL = "skill"
    ABILITY = "ability"


class CardRarity(str, Enum):
    COMMON = "common"
    UNCOMMON = "uncommon"
    RARE = "rare"


class EnemyIntentType(str, Enum):
    ATTACK = "attack"
    DEFEND = "defend"
    BUFF = "buff"
    DEBUFF = "debuff"
    UNKNOWN = "unknown"


class GamePhase(str, Enum):
    MAP = "map"
    BATTLE = "battle"
    REWARD = "reward"
    SHOP = "shop"
    REST = "rest"
    EVENT = "event"
    VICTORY = "victory"
    GAME_OVER = "game_over"


class RoomType(str, Enum):
    BATTLE = "battle"
    ELITE = "elite"
    BOSS = "boss"
    REST = "rest"
    SHOP = "shop"
    EVENT = "event"
    UNKNOWN = "unknown"


class CardEffect(BaseModel):
    type: str
    value: Optional[int] = None
    target: Optional[str] = None


class Card(BaseModel):
    id: str
    name: str
    description: str
    type: CardType
    rarity: CardRarity
    cost: int
    damage: Optional[int] = None
    block: Optional[int] = None
    effects: Optional[List[CardEffect]] = None
    upgraded: bool = False


class Buff(BaseModel):
    id: str
    name: str
    stacks: int


class Debuff(BaseModel):
    id: str
    name: str
    stacks: int


class EnemyIntent(BaseModel):
    type: EnemyIntentType
    value: Optional[int] = None


class Enemy(BaseModel):
    id: str
    name: str
    hp: int
    max_hp: int
    block: int = 0
    intent: EnemyIntent
    buffs: List[Buff] = []
    debuffs: List[Debuff] = []


class Player(BaseModel):
    hp: int
    max_hp: int
    block: int = 0
    energy: int
    max_energy: int
    gold: int
    buffs: List[Buff] = []
    debuffs: List[Debuff] = []


class Deck(BaseModel):
    draw_pile: List[Card]
    hand: List[Card]
    discard_pile: List[Card]
    exhaust_pile: List[Card] = []


class MapRoom(BaseModel):
    id: str
    type: RoomType
    completed: bool = False
    connections: List[str] = []


class MapFloor(BaseModel):
    rooms: List[MapRoom]


class GameMap(BaseModel):
    floors: List[MapFloor]


class GameState(BaseModel):
    game_id: str
    phase: GamePhase
    turn: int
    player: Player
    enemies: List[Enemy] = []
    deck: Deck
    map: GameMap
    current_floor: int
    current_room: int
    rng_seed: str
    pending_rewards: List[Card] = []


class GameStartRequest(BaseModel):
    character: Optional[str] = "ironclad"
    seed: Optional[str] = None
    ascension: Optional[int] = 0


class GameStartResponse(BaseModel):
    game_id: str
    game_state: GameState


class GameAction(BaseModel):
    type: str
    payload: Optional[dict] = None


class RewardCard(BaseModel):
    card: Card
    selected: bool = False


class RewardOffer(BaseModel):
    cards: List[RewardCard]
    gold_reward: int = 0
    can_skip: bool = True


class SelectRewardRequest(BaseModel):
    card_index: Optional[int] = None
    skip: bool = False


class RestActionRequest(BaseModel):
    action_type: str  # 'heal' or 'upgrade'


class ShopItem(BaseModel):
    card: Card
    price: int
    item_type: str  # 'card_add' or 'card_remove'


class ShopOffer(BaseModel):
    items: List[ShopItem]
    remove_price: int = 50


class ShopActionRequest(BaseModel):
    item_index: int
    action: str  # 'buy' or 'remove'
